### PR TITLE
feat(trpc): add rxjs observable compatible trpc client #379

### PR DESCRIPTION
**Pull request by Dafnik**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/analogjs/analog/blob/main/CONTRIBUTING.md#-commit-message-guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Which package are you modifying?

- [ ] vite-plugin-angular
- [ ] vite-plugin-nitro
- [ ] astro-angular
- [ ] create-analog
- [ ] router
- [ ] platform
- [ ] content
- [ ] nx-plugin
- [X] trpc

## What is the current behavior?

Current trpc client only works with `Promises`

Issue Number: #379

## What is the new behavior?

This new client works with RxJS Observables.

## Does this PR introduce a breaking change?
- [ ] Yes
- [X] No

### CODE REVIEW

**Comment by goetzrobin** [comment]
Wow thanks so much for adding all that!

I would almost say that instead of creating a new separate tRPC client and e2e project you can simply replace the existing Promise based client with the Observable one. I know this is a breaking change, but I think we are early enough to introduce it.

I think it is worth it, because the observable client fits so much better into the Angular ecosystem.

I also know that this would mean we'd have to adjust the templates of the nx-plugin, but that should be pretty straight forward. (I can obviously help with that too!)

Open to discuss of course. @brandonroberts @Dafnik what do you think?

**Comment by goetzrobin** [comment]
@brandonroberts @Dafnik  Another thing that might be worth considering to add in the scope of this is adding support for authentication. The issue I created for that is here: https://github.com/analogjs/analog/issues/394 
I think the low hanging fruit is just allow the tRPC client to take in all params that httpBatchLink takes.

**Comment by brandonroberts** [comment]
I agree. Let's land the observable compatible version first as signals are still in developer preview. Then we can revisit signal support later. Same with adding auth in a separate PR.


Thanks for working on this! 

@goetzrobin if you're good on the review, lets move to merge

**Comment by goetzrobin** [comment]
@brandonroberts we will replace the current promise based implementation with @Dafnik 's RxJs version. I will review the PR once he pushes the new changes and merge them once everything looks good 👍

**Review by goetzrobin — CHANGES_REQUESTED** [review]
Awesome work! I left a couple comments, mostly just formatting stuff. 

The thing about transferStateLink not being used is a bit of a headscratcher for me right now, but I think we could live without the caching for now and revisit in a new PR, unless it's an easy fix for you...

Thanks so much for working on this! This will be an awesome addition!

File: apps/trpc-app/src/app/app.component.ts [hunk 1]
```diff
@@ -1,10 +1,13 @@
-import { ChangeDetectionStrategy, Component } from '@angular/core';
-import { RouterOutlet } from '@angular/router';
+import { ChangeDetectionStrategy, Component } from "@angular/core";
+import { RouterOutlet } from "@angular/router";
```
[the remark this answers is not in the document]
Comment (Dafnik) [reply apps/trpc-app/src/app/app.component.ts:2]: Hey, sorry about all the importing errors. I actually run prettier before committing but I think my IDE messed with them. I will definitely fix this.

File: packages/trpc/src/lib/client/client.ts [hunk 2]
```diff
@@ -1,28 +1,20 @@
...
+import { AnyRouter } from "@trpc/server";
+import { transferStateLink } from "./links/transfer-state-link";
+import { provideTrpcCacheState, provideTrpcCacheStateStatusManager, tRPC_CACHE_STATE } from "./cache-state";
+import { createTRPCRxJSProxyClient } from "./trpc-rxjs-proxy";
+import { CreateTRPCClientOptions } from "@trpc/client/src/createTRPCUntypedClient";
 
 export type TrpcOptions<T extends AnyRouter> = {
   url: string;
-  options?: Partial<CreateTRPCClientOptions<T>>;
+  options: Partial<CreateTRPCClientOptions<T>>;
```
[the remark this answers is not in the document]
Comment (Dafnik) [reply packages/trpc/src/lib/client/client.ts:12]: Nop, just a copying error.

File: packages/trpc/src/lib/client/shared-internal.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
+import {
+  AnyRouter,
+  ClientDataTransformerOptions,
+  CombinedDataTransformer,
+  DataTransformerOptions,
+  DefaultDataTransformer
+} from "@trpc/server";
```
[the remark this answers is not in the document]
Comment (Dafnik) [reply packages/trpc/src/lib/client/shared-internal.ts:7]: This is actually I would like to work on next.

There is some tooling stuff in this repository I'd like to improve. Like migrating to pnpm, some githooks which could be simplified and some additional and updated actions to check for formatting etc.

File: apps/trpc-app/src/app/pages/index.page.ts [hunk 4]
```diff
@@ -1,401 +1,106 @@
...
-    </main>
+      </div>
+    </ng-template>
   `,
 })
 export default class HomeComponent {
   private _trpc = injectTRPCClient();
-  public loadingPosts = false;
-  public notes: Note[] = [];
+  public triggerRefresh$ = new BehaviorSubject(true);
```
[the remark this answers is not in the document]
Comment (Dafnik) [reply apps/trpc-app/src/app/pages/index.page.ts:96]: Hi, I've also just looked into this. I've added a log line in line `52` of `transfer-link-state.ts`. the `trpc.note.list.query()` is run two times and on the second run `isCacheActive.getValue()` is `false`.

Thanks for the link!

File: packages/trpc/src/lib/client/client.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Dafnik) [reply packages/trpc/src/lib/client/client.ts]: Yeah, I was working on the other stuff too. haha but thx

File: apps/trpc-app/src/app/pages/index.page.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Dafnik) [reply apps/trpc-app/src/app/pages/index.page.ts:96]: You are right, it looks like it has something to do with the transfer cache active subject.

[the remark this answers is not in the document]
Comment (goetzrobin) [reply apps/trpc-app/src/app/pages/index.page.ts:96]: I vaguely remember there being a reason why I did it that way, but I am not completely sure anymore...

I think the subscription through `notes$ | async` in the template might be after the app already stabilized and therefore is after the cache is active

**Review by goetzrobin — APPROVED** [review]

File: apps/trpc-app/src/app/pages/index.page.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Dafnik) [reply apps/trpc-app/src/app/pages/index.page.ts:96]: Hi, yeah, everything was already done! Just didn't get to reply you yet but thank you for reviewing and merging everything!

**Comment by Dafnik** [comment]
It was my pleasure @goetzrobin! Thanks for doing all of this work and I would love to also help out in the future.

**Pull request merged**